src/main/scala/li/cil/oc/server/component/UpgradeSkin.scala

```diff
@@ -0,0 +1,334 @@
+        } else if(stack.stackSize == 0) {
+          inventory.setInventorySlotContents(selectedSlot, skin.get)
+          stack.stackSize += 1
+        } else {
+          stack.stackSize += 1
+        }
+        changed = true
+        skin = None
+      }
+
+      if(!onlyDeposit) {
+        if (stack != null && skin.isEmpty && api.Items.get(extracted.get) == api.Items.get(Constants.BlockName.Print)) {
+          skin = extracted
+          changed = true
+        } else if (stack != null) {
+          stack.stackSize += 1
+        }
+
+        if (stack != null && stack.stackSize == 0) {
+          inventory.setInventorySlotContents(selectedSlot, null)
+        }
+      }
+      skinComponent.skin = skin
+      inventory.markDirty()
+      host.markDirty()
+      updateClient()
+      changed
+    }
+
+    def setSkinRotation(skinComponent: UpgradeSkinComponent, args: Arguments): Array[AnyRef] = {
+      var rotX = 0f
+      var rotY = 0f
+      var rotZ = 0f
+      if(args.isInteger(1)) {
+        rotX = args.checkInteger(1)
+      } else if(args.isDouble(1)) {
+        rotX = args.checkDouble(1).toFloat
+      } else {
+        return result(Unit, "Number expected: rotX")
+      }
+      if(args.isInteger(2)) {
+        rotY = args.checkInteger(2)
+      } else if(args.isDouble(2)) {
+        rotY = args.checkDouble(2).toFloat
+      } else {
+        return result(Unit, "Number expected: rotY")
+      }
+      if(args.isInteger(3)) {
+        rotZ = args.checkInteger(3)
+      } else if(args.isDouble(3)) {
+        rotZ = args.checkDouble(3).toFloat
+      } else {
+        return result(Unit, "Number expected: rotZ")
+      }
+
+      if(skinComponent.animationTicksLeft == 0) {
+        val maxRot = math.max(math.max(math.abs(rotX - skinComponent.oldRotX), math.abs(rotY - skinComponent.oldRotY)), math.abs(rotZ - skinComponent.oldRotZ))
+        val moveTicks = math.max((Settings.get.turnDelay * 20).toInt, 1) // normalize the duration the rotation takes with the distance the rotation takes (90deg = turnDelay * 20)
+        skinComponent.prepareRotationAnimation(rotX, rotY, rotZ, moveTicks)
+        host.markDirty()
+        updateClient()
+        result(true)
+      } else {
+        result(Unit, "Skin is already rotating!")
+      }
+    }
+
+    def hideSomething(apply: Boolean => Unit, args: Arguments): Array[AnyRef] = {
+      if(args.isBoolean(0)) {
+        apply(args.checkBoolean(0))
+      } else {
+        apply(true)
+      }
+      host.markDirty()
+      updateClient()
+      result(true)
+    }
+
+    private def updateClient(): Unit = host match {
+      case robot: internal.Robot => robot.synchronizeSlot(robot.componentSlot(node.address))
+      case _ =>
+    }
+
+    def checkIfTierIsHighEnoughForPart(part: Int): Boolean = {
+      ((part == 2 || part == 5) && tier >= 1) || ((part == 3 || part == 6) && tier >= 2) || (part == 0 || part == 1 || part == 4)
```

Comment:
This would probably be much more readable if we had an Enum for the part number; a converter could try to match the integer value to an Enum, or emit an error if it is not a valid part index.

Reply:
Yeah, I thought about that. The design consideration here was, that you use these numbers in the ingame Lua programs, so it should be referenced as numbers in the code aswell.
But if you disagree, I'll change it asap.

Reply:
An user can always create their own constants for such numeric values in Lua code - for example, OpenOS comes with `/lib/sides.lua` and `/lib/colors.lua`. I think we should do the same in Scala code; the Enum will document what the part numbers are by virtue of the ordinal.